src/main/java/com/iexec/worker/compute/pre/PreComputeService.java

```diff
@@ -129,12 +133,17 @@ public PreComputeResponse runTeePreCompute(TaskDescription taskDescription, Work
             log.info("Task contains TEE input data [chainTaskId:{}, containsDataset:{}, " +
                             "containsInputFiles:{}]", chainTaskId, taskDescription.containsDataset(),
                     taskDescription.containsInputFiles());
-            Integer exitCode = prepareTeeInputData(taskDescription, secureSessionId);
-            if (exitCode == null || exitCode != 0) {
-                ReplicateStatusCause exitCause = getExitCause(chainTaskId, exitCode);
-                log.error("Failed to prepare TEE input data [chainTaskId:{}, " +
-                        "exitCode:{}, exitCause:{}]", chainTaskId, exitCode, exitCause);
-                preComputeResponse.setExitCause(exitCause);
+            try {
+                Integer exitCode = prepareTeeInputData(taskDescription, secureSessionId);
+                if (exitCode == null || exitCode != 0) {
+                    ReplicateStatusCause exitCause = getExitCause(chainTaskId, exitCode);
+                    log.error("Failed to prepare TEE input data [chainTaskId:{}, " +
+                            "exitCode:{}, exitCause:{}]", chainTaskId, exitCode, exitCause);
+                    preComputeResponse.setExitCause(exitCause);
+                    return preComputeResponse;
+                }
+            } catch (TimeoutException e) {
+                preComputeResponse.setExitCause(PRE_COMPUTE_TIMEOUT);
                 return preComputeResponse;
             }
```

Comment:
That wouldn't be correct. If `exitCode == null || exitCode != 0` is false, then we do not want to return the `preComputeResponse` before setting the `secureSessionId` field.

Reply:
As we have imporoved `PrecomputeResponse isSuccessful` implementation, the code should be updatable now.
* set `secureSessionId` before downloading dataset
* set `exitCause` as needed (timeout or other cause)
* have a single return `PreComputeResponse` instance at the end of the execution

What do you think ? Did I miss something ?

Reply:
OK for 1 & 2. 3 is a bit harder to do so I'll keep the two first `return` statements.

Reply:
Thanks for e31f521. I have further questions around the logic behind the new implementation.
If I understand correctly, `PreComputeResponse` will have an explicit `FAILED` `finalStatus` in 2 cases
* bad enclave configuration
* bad enclave heap size

Then, if TEE session creation fails, a PreComputeResponse will be returned without defining the final status.

Otherwise, the pre compute execution will be executed.
If there is an error, we retrieve it with the `exitCause`.
If the `exitCause` is `PRE_COMPUTE_TIMEOUT`, there is no final status defined for PreComputeResponse (we don't use `DockerRunFinalStatus.TIMEOUT`)

After pre-compute execution, if there is no `exitCause` and no TEE session error happened, we set the finalStatus to `DockerRunFinalStatus.SUCCESS`.

I think I got your point but I don't really feel safe with the cases of is `finalStatus` defined or not ?
The questions to answer for me are:
* Do we need both exitCause and finalStatus in PreComputeResponse ?
* If we need them both, is there one which is more important than the other ?
* Could they be merged with safety checks against null values ?

Reply:
You're right, this implementation is misleading. Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/455/commits/0ab3b0dbd28086ee8056e869bf7a4e8b70bcc60c. Its goal is to simplify the error handling and make the `Pre/App/PostComputeResponse` more similar and easier to handle.

Reply:
Seems a lot less confusing to me right now !
Great work !